daemon/cdi_linux.go

```diff
@@ -0,0 +1,85 @@
+package daemon
+
+import (
+	"fmt"
+
+	"github.com/container-orchestrated-devices/container-device-interface/pkg/cdi"
+	"github.com/docker/docker/pkg/capabilities"
+	"github.com/hashicorp/go-multierror"
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/sirupsen/logrus"
+)
+
+// cdiDeviceInjector defines the operations required to inject CDI devices into an OCI spec.
+type cdiDeviceInjector interface {
+	InjectDevices(*specs.Spec, ...string) ([]string, error)
+	GetErrors() map[string][]error
+}
+
+type cdiHandler struct {
+	registry cdiDeviceInjector
+}
+
+// RegisterCDIDriver registers the CDI device driver.
+// The driver injects CDI devices into an incoming OCI spec and is called for DeviceRequests associated with CDI devices.
+func RegisterCDIDriver(opts ...cdi.Option) {
+	var registry cdiDeviceInjector
+
+	cache, err := cdi.NewCache(opts...)
+	if err != nil {
+		// We store an invalid cache that always returns an error.
+		// This error will be returned when a CDI device is requested. This ensures that daemon startup is not blocked
+		// by a CDI registry initialization failure.
+		registry = &invalidCache{err: fmt.Errorf("failed to initialize CDI registry: %w", err)}
```

Comment:
This is way more complex than necessary. Quit trying to fit an error-shaped peg into a `cdiHandler`-shaped hole and simply register the `deviceDriver` with an `updateSpec` function which returns the error.

Reply:
Thanks. That's a great suggestion. Will update.